### PR TITLE
Revisions SDL-0245 WiFi Transport: Sharing SSID and password with Mobile Proxy

### PR DESCRIPTION
**Pull request by ashwink11**
Changes added as per review comments:

>* Added WiFi AP and station mode support for IVI.
>* Updated mobile APIs, HMI use cases

### CODE REVIEW

**Comment by ashwink11** [comment]
@theresalech Please review this PR.

File: proposals/0245-sharing-wifi-ssid-and-password.md [hunk 1]
```diff
@@ -288,102 +508,101 @@ Hotspot Configuration
...
-2. User tries to launch the app on HMI.
-3. HMI determines that WiFi auto-connect is supported using `DeviceInfo.supportWiFiAutoConnect`.
+1. If the proxy uses WiFi mobile APIs defined above before user consent, SDL Core responds with `DISALLOWED`.
+2. The user tries to launch the app on HMI.
+3. HMI determines that if WiFi connection can be automatically established using `DeviceInfo.wifiConfigurationInfo`.
 4. HMI requests `GetListOfPermissions` and determines if `WiFi group` permissions needed.
-5. HMI requests user consent to share WiFi credentials with app and informs the user that credentials can be used by the app to connect WiFi automatically.
-6. The user provides consent to share WiFi credentials with app. 
-7. HMI shows `Waiting for WiFi Connection. Please check your device when it is safe to do so. Apps on some devices may need your permissions to connect Car WiFi.` pop-up with a timeout of 5 seconds.
+5. HMI requests user consent to share WiFi credentials with the app and informs the user that credentials can be used the app or IVI to join the WiFi network automatically.
```
Comment (theresalech) [inline proposals/0245-sharing-wifi-ssid-and-password.md]:
```suggestion
5. HMI requests user consent to share WiFi credentials with the app and informs the user that credentials can be used by the app or IVI to join the WiFi network automatically.
```

File: proposals/0245-sharing-wifi-ssid-and-password.md [hunk 2]
```diff
@@ -288,102 +508,101 @@ Hotspot Configuration
...
+7. The IVI shows `Waiting for WiFi Connection. Please check your device when it is safe to do so. Apps on some devices may need your permissions to join WiFi network.` pop-up with a timeout of 5 seconds.
 
 ##### Post Conditions 
 
-1. The proxy receives `OnPermissionChange` notification. As soon as this notification is received, the proxy requests WiFi credentials using `GetWiFiStatusInfo` Mobile API. 
-2. HMI sends `OnSystemRequest` notification with `CONNECT_WIFI` request type.
-3. WiFi credentials shared with the proxy.
-4. The proxy should try connecting to WiFi as soon as possible on receiving `OnSystemRequest` notification with `CONNECT_WIFI` request type.
+1. The proxy receives `OnPermissionChange` notification.
+2. HMI sends `OnSystemRequest` notification with `WIFI_CONNECTION_REQ` request type. On receiving `OnSystemRequest` notification with `WIFI_CONNECTION_REQ` request type, the proxy should use mobile APIs for wifi to get or share wifi credentials.
```
Comment (theresalech) [inline proposals/0245-sharing-wifi-ssid-and-password.md]:
```suggestion
2. HMI sends `OnSystemRequest` notification with `WIFI_CONNECTION_REQ` request type. On receiving `OnSystemRequest` notification with `WIFI_CONNECTION_REQ` request type, the proxy should use mobile APIs for WiFi to get or share WiFi credentials.
```

File: proposals/0245-sharing-wifi-ssid-and-password.md [hunk 3]
```diff
@@ -288,102 +508,101 @@ Hotspot Configuration
...
 
 ##### Scenario Description
-1. If the proxy requests `WiFi` status info, SDL Core responds with `SUCCESS` and provides WiFi credentials to app.
-2. User tries to launch the app on HMI.
-3. HMI sends `OnSystemRequest` notification with `CONNECT_WIFI` request type.
-4. HMI shows `Waiting for WiFi Connection. Please check your device when it is safe to do so. Some devices may need your permissions to connect Car WiFi.` pop-up with a timeout of 5 seconds.
+1. The user tries to launch the app on HMI.
+2. HMI sends `OnSystemRequest` notification with `WIFI_CONNECTION_REQ` request type.
+3. HMI shows `Waiting for WiFi Connection. Please check your device when it is safe to do so. Some devices may need your permissions to connect Car WiFi.` pop-up with a timeout of 5 seconds.
+4. If IVI is a WiFi access point, the proxy use `GetWiFiNetworkInfo` RPC to requests WiFi status info, the SDL Core responds with `SUCCESS` and provides WiFi credentials to the app.
```
Comment (theresalech) [inline proposals/0245-sharing-wifi-ssid-and-password.md]:
```suggestion
4. If IVI is a WiFi access point, the proxy uses `GetWiFiNetworkInfo` RPC to request WiFi status info, then SDL Core responds with `SUCCESS` and provides WiFi credentials to the app.
```

File: proposals/0245-sharing-wifi-ssid-and-password.md [hunk 4]
```diff
@@ -288,102 +508,101 @@ Hotspot Configuration
...
+5. If the mobile device is a WiFi access point, the proxy use `JoinWiFiNetwork` RPC to share WiFi status info with the IVI, the SDL Core will forward WiFi info using HMI API `JoinWiFiNetwork` to HMI and receive result code. The SDL Core responds with `SUCCESS` on a successful WiFi connection.
```
Comment (theresalech) [inline proposals/0245-sharing-wifi-ssid-and-password.md]:
```suggestion
5. If the mobile device is a WiFi access point, the proxy uses `JoinWiFiNetwork` RPC to share WiFi status info with the IVI, then SDL Core will forward WiFi info using HMI API `JoinWiFiNetwork` to HMI and receive result code. SDL Core responds with `SUCCESS` on a successful WiFi connection.
```

File: proposals/0245-sharing-wifi-ssid-and-password.md [hunk 5]
```diff
@@ -392,27 +611,27 @@ This case is applicable when driver distraction is ON.
...
+1. The user tries to launch the app on HMI.
 2. HMI shows `Connect WiFi to use this App` pop-up with a timeout of 5 seconds.
 
 ##### Post Conditions 
 
-1. User manually connects WiFi to use app.
+1. The user manually connects WiFi to use the app.
 
-#### **E. WiFi auto-connect feature not enabled for app**
+#### **E. Automatice WiFi connection cannot be established**
```
Comment (theresalech) [inline proposals/0245-sharing-wifi-ssid-and-password.md]:
```suggestion
#### **E. Automatic WiFi connection cannot be established**
```

File: proposals/0245-sharing-wifi-ssid-and-password.md [hunk 6]
```diff
@@ -7,59 +7,171 @@
...
+---
+
+The core will know the supported secondary transport based on values in the `smartdevicelink.ini` file. The possible values for the supported secondary transport value need to be updated. `WiFi_AP_MODE` and `WiFi_STA_MODE` is added below for supported secondary transport values. The SDL core will use these values to communicate supported secondary transport using protocol message `Start Service Ack`.
+
+The SDL adapters can define desired supported secondary transports in `smartdevicelink.ini` file as shown below.
+
+```
+; Comma-separated list of transports that can be used as Secondary Transport for each Primary Transport.
+; Possible values are WiFi, WiFi_AP_MODE, WiFi_STA_MODE, USB and Bluetooth.
+; Core will not suggest Secondary Transport if the value is empty.
```
[the remark this answers is not in the document]
Comment (ashwink11) [reply proposals/0245-sharing-wifi-ssid-and-password.md]: All comments in the smartdevicelink.ini file are preceded by semi-colons. I just updated the text in those comments. Reference: https://github.com/smartdevicelink/sdl_core/blob/master/src/appMain/smartDeviceLink.ini

File: proposals/0245-sharing-wifi-ssid-and-password.md [hunk 7]
```diff
@@ -7,59 +7,171 @@
...
+; in preferred order. If a transport is not listed, then the service is not allowed
+; to run on the transport.
+; Only video and audio services are configurable.
+; If the entry of a service is completely omitted, the service will be allowed on all transports.
+; Possible values are: IAP_BLUETOOTH, IAP_USB, IAP_USB_HOST_MODE, IAP_USB_DEVICE_MODE, IAP_CARPLAY, SPP_BLUETOOTH, AOA_USB,  TCP_WIFI, TCP_WIFI_AP_MODE and TCP_WIFI_STA_MODE.
+; Note: this configuration is applied even if multiple-transports feature is not enabled.
+; Audio service
+0x0A = TCP_WIFI_STA_MODE, IAP_CARPLAY, IAP_USB_HOST_MODE, IAP_USB_DEVICE_MODE, IAP_USB, AOA_USB
+; Video service
+0x0B = TCP_WIFI_STA_MODE, IAP_CARPLAY, IAP_USB_HOST_MODE, IAP_USB_DEVICE_MODE, IAP_USB, AOA_USB
```
[the remark this answers is not in the document]
Comment (ashwink11) [reply proposals/0245-sharing-wifi-ssid-and-password.md]: All comments in the smartdevicelink.ini file are preceded by semi-colons. I just updated the text in those comments. Reference: https://github.com/smartdevicelink/sdl_core/blob/master/src/appMain/smartDeviceLink.ini

File: proposals/0245-sharing-wifi-ssid-and-password.md [hunk 8]
```diff
@@ -7,59 +7,171 @@
...
+    <param name="password" type="String" minlength="1" maxlength="100" mandatory="false">
+        <description>password to use to connect AP</description>
+    </param>
+    <param name="securityType" type="WiFiSecurityType" mandatory="false">
+        <description>security type of WiFi AP</description>
+    </param>   
+</function>
+
+<function name="JoinWiFiNetwork" messagetype="response">
+</function>
```
[the remark this answers is not in the document]
Comment (ashwink11) [reply proposals/0245-sharing-wifi-ssid-and-password.md]: Added

**Comment by ashwink11** [comment]
@theresalech added changes as per review comments.

File: proposals/0245-sharing-wifi-ssid-and-password.md [hunk 9]
```diff
@@ -194,20 +319,61 @@ Enum to define WiFi state and WiFi security type.
...
 ```
 
 Possible responses for the above mentioned RPCs.
 
 1. RPC response will be `SUCCESS` if the request is successful.
-2. RPC response will be `UNSUPPORTED_RESOURCE` if HMI does not support WiFi transport for SDL.
+2. RPC response will be `UNSUPPORTED_RESOURCE` if HMI does not support WiFi transport for SDL. The SDL Core will also send this result code in the following scenarios:
+  >* If `JoinWiFiNetwork` RPC is used, when supported secondary transport received by the mobile app in `Start Service Ack` is only `TCP_WIFI_STA_MODE`.
+  >* If `GetWiFiNetworkInfo` RPC is used, when supported secondary transport received by the mobile app in `Start Service Ack`  is only `TCP_WIFI_AP_MODE`.
```
[the remark this answers is not in the document]
Comment (ashwink11) [reply proposals/0245-sharing-wifi-ssid-and-password.md]: Done.

File: proposals/0245-sharing-wifi-ssid-and-password.md [hunk 10]
```diff
@@ -7,59 +7,171 @@
...
+1. Device Info is changed to provide the apps with WiFi capabilities to IVI. IVI will receive `DeviceInfo.wifiConfigurationInfo` in `OnAppRegistered` HMI API when the app is registered using `primary transport`. 
+
+>* If the app can connect automatically to the WiFi access point created by the IVI then the `DeviceInfo.wifiConfigurationInfo.autoJoinWiFiSupported` will be `true`.
+>* If the app supports WiFi access point mode then the `DeviceInfo.wifiConfigurationInfo.canHostWiFiNetwork` will be `true`.
+
+2. HMI can send `OnSystemRequest` notification with `RequestType` as `WIFI_CONNECTION_REQ` to inform the proxy that it can initiate an automatic WiFi connection. This notification will be sent only when a user has provided consent to share WiFi credentials, and the connected app has supported WiFi configuration and the user attempts to launch that app. 
+3. If WiFi configurations of the IVI and the app do not match, they cannot automatically join the WiFi network. Hence, the IVI should not send `OnSystemRequest` notification with `RequestType` as `WIFI_CONNECTION_REQ`.
```
[the remark this answers is not in the document]
Comment (ashwink11) [reply proposals/0245-sharing-wifi-ssid-and-password.md]: Done.

File: proposals/0245-sharing-wifi-ssid-and-password.md [hunk 11]
```diff
@@ -216,31 +382,64 @@ Apps need to include additional permissions to use APIs to connect to WiFi. Howe
...
+
+Here is an example of the WiFi configuration APIs usage to set `DeviceInfo.wifiConfigurationInfo` in Android proxy.
+
+>* If `MultiplexTransportConfig.setCanHostWiFiNetwork( true )` and mobile device __DOES NOT__ supports WiFi transport APIs provided by platform SDK then `DeviceInfo.wifiConfigurationInfo.canHostWiFiNetwork = false`
+---
+>* If `MultiplexTransportConfig.setCanAutoJoinWiFiNetwork( true )` and mobile device __DOES NOT__ supports WiFi transport APIs provided by platform SDK then `DeviceInfo.wifiConfigurationInfo.canAutoJoinWiFiNetwork = false`
+---
+>* If `MultiplexTransportConfig.setCanHostWiFiNetwork( false )` and mobile device __DOES__ supports WiFi transport APIs provided by platform SDK then `DeviceInfo.wifiConfigurationInfo.canHostWiFiNetwork = false`
+---
+>* If `MultiplexTransportConfig.setCanAutoJoinWiFiNetwork( false )` and mobile device __DOES__ supports WiFi transport APIs provided by platform SDK then `DeviceInfo.wifiConfigurationInfo.canAutoJoinWiFiNetwork= false`
```
[the remark this answers is not in the document]
Comment (ashwink11) [reply proposals/0245-sharing-wifi-ssid-and-password.md]: Done.

**Pull request merged**